torch/lib/c10d/ProcessGroupWrapper.cpp

```diff
@@ -0,0 +1,314 @@
+    std::vector<at::Tensor> inp_tensors({tensor});
+    pg->allgather(out_tensors, inp_tensors)->wait();
+    std::unordered_set<int64_t> gathered;
+    for (const auto& t : out_tensors[0]) {
+      auto n = t.item().to<int64_t>();
+      gathered.insert(n);
+      if (gathered.size() > 1) {
+        TORCH_CHECK(false, failureMsg);
+      }
+    }
+  }
+
+  // Verifies that shapes are consistent across processes.
+  // shape_tensors_to_report should be specified as the tensors to report when a
+  // shape inconsistency is found. This is not necessarily shape_tensors such as
+  // in the case we are checking shape dimensionality.
+  void verify_shapes(
+      std::vector<at::Tensor> shape_tensors,
+      std::vector<at::Tensor> shape_tensors_to_report,
+      c10::intrusive_ptr<ProcessGroup>& pg) {
+    std::vector<std::vector<at::Tensor>> output_tensors;
+    output_tensors.reserve(shape_tensors.size());
+    for (int k = 0; k < shape_tensors.size(); ++k) {
+      auto& tensor_shape = shape_tensors[k];
+      std::vector<at::Tensor> outputs;
+      outputs.reserve(pg->getSize());
+      for (int i = 0; i < pg->getSize(); ++i) {
+        outputs.emplace_back(at::zeros_like(tensor_shape));
+      }
+      output_tensors.emplace_back(outputs);
+    }
+    // Allgather tensor shapes.
+    pg->allgather(output_tensors, shape_tensors)->wait();
+    // Verify equivalence
+    for (int i = 0; i < output_tensors.size(); ++i) {
+      auto world_tensor_shapes = output_tensors[i];
+      auto reference_shape_tensor = shape_tensors[i];
+      for (const auto& rank_tensor_shape : world_tensor_shapes) {
+        if (!rank_tensor_shape.equal(reference_shape_tensor)) {
+          TORCH_CHECK(
+              false,
+              c10::str(
+                  "Error when verifying shape tensors for collective ",
+                  opTypeToString(op_type_),
+                  " on rank ",
+                  pg->getRank(),
+                  ". This likely indicates that input shapes into the collective are mismatched across ranks. Got shapes: ",
+                  shape_tensors_to_report));
+        }
+      }
+    }
+  }
+
+  // Executes and verifies the collective fingerprint.
+  void verify(c10::intrusive_ptr<ProcessGroup> pg) {
+    // For collectives, all ranks should participate and call into them in the
+    // same order. Verify the same operation type is being requested.
+    int64_t op_type_int = static_cast<int64_t>(op_type_);
+    verify_num(
+        op_type_int,
+        pg,
+        c10::str(
+            "Mismatch between collective operation types across ranks.",
+            "This likely indicates an application bug where different ranks are ",
+            "calling different collectives. ",
+            "Rank ",
+            pg->getRank(),
+            " is calling collective: ",
+            opTypeToString(op_type_)));
+    if (input_tensors_.size() == 0) {
+      // No shapes to verify.
+      return;
+    }
```

Comment:
You're right, this can happen if one rank does allreduce with a tensor and the other rank somehow does `pg.allreduce([])`. Although most collectives seem to do this sort of input checking where they throw earlier if the input is empty (e.g. `allreduce` does this).

Still, we don't need to have this and can remove it, letting the shape checking occur and error out if it is inconsistent.